### PR TITLE
Read only necessary columns & respect `ttl_only_drop_parts` on `materialize ttl`

### PR DESCRIPTION
**Pull request by zvonand**
Finish #59189. A step in direction of implementing #20451.

Respect `ttl_only_drop_parts` on `materialize ttl`. Only read the necessary columns to recalculate TTL and drop parts by replacing them with an empty one.

### Changelog category (leave one):
- Performance Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Respect `ttl_only_drop_parts` on `materialize ttl`; only read necessary columns to recalculate TTL and drop parts by replacing them with an empty one.

### Documentation entry for user-facing changes

- [ ] Documentation is written (mandatory for new features)

> Information about CI checks: https://clickhouse.com/docs/en/development/continuous-integration/

#### CI Settings (Only check the boxes if you know what you are doing):
- [ ]  Allow: All Required Checks
- [ ]  Allow: Stateless tests
- [ ]  Allow: Stateful tests
- [ ]  Allow: Integration Tests
- [ ]  Allow: Performance tests
- [ ]  Allow: All Builds
- [ ]  Allow: batch 1, 2 for multi-batch jobs
- [ ]  Allow: batch 3, 4, 5, 6 for multi-batch jobs
---
- [ ]  Exclude: Style check
- [ ]  Exclude: Fast test
- [ ]  Exclude: All with ASAN
- [ ]  Exclude: All with TSAN, MSAN, UBSAN, Coverage
- [ ]  Exclude: All with aarch64, release, debug
---
- [ ]  Do not test
- [ ]  Woolen Wolfdog
- [ ]  Upload binaries for special builds
- [ ]  Disable merge-commit
- [ ]  Disable CI cache

### CODE REVIEW

**Comment by robot-ch-test-poll1** [comment]
*This is an automated comment for commit a6a2997a3a048a1aa25530e391ad555dcba46c04 with description of existing statuses. It's updated for the latest CI running*

[⏳ Click here](https://s3.amazonaws.com/clickhouse-test-reports/65488/a6a2997a3a048a1aa25530e391ad555dcba46c04/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Cloud fork sync (only for ClickHouse Inc. employees)</td><td>If it fails, ask a maintainer for help</td><td>⏳ pending</td></tr>
<tbody>
</table>
<details><summary>Successful checks</summary>
<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>PR Check</td><td>Checks correctness of the PR's body</td><td>✅ success</td></tr>
<tbody>
</table>
</details>

**Comment by zvonand** [comment]
Fails:

* `Integration tests (aarch64) [5/6]`
  - `test_when_s3_broken_pipe_at_upload_is_retried` -- looks like #61457
* `Stateless tests (aarch64)`
  - `02125_many_mutations` -- #65589
* `Stateless tests (ubsan) [2/2]`
  - `02922_deduplication_with_zero_copy` -- fixed by #65492

**Comment by alexey-milovidov** [comment]
@zvonand, waiting to fix all the flaky tests.

**Comment by zvonand** [comment]
@alexey-milovidov you mean `02125_many_mutations` ?

**Comment by zvonand** [comment]
> @zvonand, waiting to fix all the flaky tests.

If you mean `02125_many_mutations`, then I took a closer look at it and I do not see a reason why the test itself shall be failing. Maybe this is caused by its nature: 20000 ALTERs is a huge number, e.g. this may interfere with something inside runner.

I cannot poke the CI from the inside, and, as soon as the flakiness is not related to this PR, I do not see why it is a stopper here. Also, this test has been failing from time to time ever since it was introduced in 2021.

**Comment by alexey-milovidov** [comment]
@zvonand A test for LDAP has failed: https://s3.amazonaws.com/clickhouse-test-reports/65562/57750540c4cbad4b3ea6719c14eb4611f67663f8/integration_tests__aarch64__[4_6].html

File: src/Storages/StorageInMemoryMetadata.h
Comment (CurtizJ) [inline src/Storages/StorageInMemoryMetadata.h:148]: A bit strange name, I'd name it `hasOnlyRowsTTL`.

**Review by CurtizJ — CHANGES_REQUESTED** [review]
In general, I think this approach won't work. I'll think how it's possible to solve this task in the other way.

**Comment by CurtizJ** [comment]
Hmm, probably it will work. I'll check one more time more closely.

**Comment by zvonand** [comment]
> Hmm, probably it will work. I'll check one more time more closely.

@CurtizJ what is your opinion now, does it work? For me, it looks like it does :)

**Comment by zvonand** [comment]
@CurtizJ so, what do you think of this now? (Also, the CI for this PR is really green)

It actually seems to work:  
1. Create table:
```
SET mutations_sync = 2;
CREATE TABLE test_ttl
(
    `ts` DateTime CODEC(Delta, LZ4),
    `huge_column1` String CODEC(NONE),
    `huge_column2` String CODEC(NONE),
    `huge_column3` String CODEC(NONE)
)
ENGINE = MergeTree
PARTITION BY toDate(ts)
ORDER BY tuple() AS
SELECT
    toDateTime('2020-01-01 00:00:00') + toIntervalMillisecond(number * 5) AS ts,
    randomPrintableASCII(100) AS huge_column1,
    randomPrintableASCII(120) AS huge_column2,
    randomPrintableASCII(130) AS huge_column3
FROM numbers(100000000);
```
2. Alter it:
```
ALTER TABLE test_ttl
    MODIFY SETTING ttl_only_drop_parts = 1;
ALTER TABLE test_ttl
    MODIFY TTL ts + toIntervalDay(1439);
```

Difference in performance is dramatic (60 seconds in master vs 1 second in this PR)

**Comment by alexey-milovidov** [comment]
@zvonand, a test for LDAP has failed in the master branch: https://s3.amazonaws.com/clickhouse-test-reports/0/c945209427e759dccfaff5a6af11f2e18c100390/integration_tests__aarch64__[4_6].html

**Comment by zvonand** [comment]
Yes, I already seen it and thinking about it, even without 3 reminders in 3 unrelated PRs.  
By the way, there is a standalone issue for this test failure: #65762  
Also, this PR does not have any tests failed, and it is totally green.

**Comment by zvonand** [comment]
@CurtizJ gentle ping

**Comment by alexey-milovidov** [comment]
@zvonand, a test for authentication by SSL certificates has failed:
https://s3.amazonaws.com/clickhouse-test-reports/0/aa96fd3385410cfddd39187669894034e3b83cc0/integration_tests__release__[4_4].html
https://s3.amazonaws.com/clickhouse-test-reports/0/2e21683bedcd97d9090955316765daae4d631065/integration_tests__aarch64__[6_6].html

**Comment by zvonand** [comment]
@alexey-milovidov do you want me to look into it? I can, but without promise of quick result.

I also see that other PRs are being reviewed and merged during the last couple of weeks, even though https://aretestsgreenyet.com/ [has never been green](https://github.com/ClickHouse/ClickHouse/pull/59138#issuecomment-2254635605). Maybe I can also have some, esp. on PRs that have been waiting for months?

P.S. Btw, looks like the test's original author was @vitlibar, maybe he would know better

**Comment by zvonand** [comment]
@alexey-milovidov none of the tests failed, but looks this page is not supposed do be green at all, right? :) 

image

**Comment by alexey-milovidov** [comment]
It requires at least one million test runs. If it is less than one million it indicates that something is wrong with the CI or everyone is on vacation.

**Comment by alexey-milovidov** [comment]
@zvonand, check this: https://s3.amazonaws.com/clickhouse-test-reports/0/0837a51313c5991e57265a5fd3556908bb6f4c2a/integration_tests__asan__old_analyzer__[6_6].html

**Comment by zvonand** [comment]
> @zvonand, check this: [https://s3.amazonaws.com/clickhouse-test-reports/0/0837a51313c5991e57265a5fd3556908bb6f4c2a/integration_tests__asan__old_analyzer__[6_6].html](https://s3.amazonaws.com/clickhouse-test-reports/0/0837a51313c5991e57265a5fd3556908bb6f4c2a/integration_tests__asan__old_analyzer__%5B6_6%5D.html)

I already fixed #66835, now I am also investigating on #68355 in free time (but looks like the test itself is not the only issue here), and I have a couple of other tests I plan to take a look at later if they are not fixed by then. I will put this on the list.  

TBH I expected some reciprocity, i.e. someone would take a look at some of my not-so-huge PRs that have been waiting for review for months (e.g. this PR or #59138) when I contribute to better tests.  
Don't get me wrong, but ghosted PRs are not a good motivation.

File: src/Storages/MergeTree/MutateTask.cpp
Comment (CurtizJ) [reply src/Storages/MergeTree/MutateTask.cpp:2384]: Please, fix this.

File: src/Storages/MergeTree/MutateTask.cpp
Comment (CurtizJ) [inline src/Storages/MergeTree/MutateTask.cpp:2429]: And here.

**Review by CurtizJ — COMMENTED** [review]
Well, I agree that it works. Please fix the requested changes and I'll merge PR. Now your code doesn't remove any parts but works only like `materialize_ttl_recalculate_only`. Parts are actually removed by assigned TTL merge.

```sql
SET mutations_sync = 2

CREATE TABLE test_ttl
(
    `ts` DateTime CODEC(Delta, LZ4),
    `huge_column1` String CODEC(NONE),
    `huge_column2` String CODEC(NONE),
    `huge_column3` String CODEC(NONE)
)
ENGINE = MergeTree
PARTITION BY toDate(ts)
ORDER BY tuple()
AS SELECT
    toDateTime('2020-01-01 00:00:00') + toIntervalMillisecond(number * 5) AS ts,
    randomPrintableASCII(100) AS huge_column1,
    randomPrintableASCII(120) AS huge_column2,
    randomPrintableASCII(130) AS huge_column3
FROM numbers(10000000);

ALTER TABLE test_ttl MODIFY SETTING ttl_only_drop_parts = 1;

ALTER TABLE test_ttl MODIFY TTL ts + toIntervalDay(1439);

SYSTEM FLUSH LOGS;

SELECT
    name,
    event_type,
    merge_reason
FROM system.part_log
WHERE (`table` = 'test_ttl') AND (CAST(splitByChar('_', part_name)[4], 'Int64') > 0) AND (event_type IN ('MergeParts', 'MutatePart'))
ORDER BY event_time_microseconds ASC

Query id: ac76af54-75c4-4bce-95d3-920c5d6bcac9

   ┌─name───────────┬─event_type─┬─merge_reason─┐
1. │ 20200101_1_6_1 │ MergeParts │ RegularMerge │
   └────────────────┴────────────┴──────────────┘
   ┌─name────────────┬─event_type─┬─merge_reason─┐
2. │ 20200101_7_12_1 │ MergeParts │ RegularMerge │
   └─────────────────┴────────────┴──────────────┘
   ┌─name───────────────┬─event_type─┬─merge_reason───┐
3. │ 20200101_7_12_1_15 │ MutatePart │ NotAMerge      │
4. │ 20200101_1_6_1_15  │ MutatePart │ NotAMerge      │
5. │ 20200101_1_14_2_15 │ MergeParts │ TTLDeleteMerge │
   └────────────────────┴────────────┴────────────────┘
```

File: src/Storages/MergeTree/MutateTask.cpp
Comment (zvonand) [reply src/Storages/MergeTree/MutateTask.cpp:2429]: Sorry, I really didn't get it. What is wrong with this and the one above?

Comment (CurtizJ) [reply src/Storages/MergeTree/MutateTask.cpp:2429]: The line `task = std::make_unique<MutateSomePartColumnsTask>(ctx);` is always called and overrides the task. Therefore your previous code with `if-else` is not effective.

**Comment by alexey-milovidov** [comment]
I've resolved the conflict in private. Let's see what the CI will show. Waiting for a final approval from @CurtizJ

**Review by CurtizJ — APPROVED** [review]

**Comment by CurtizJ** [comment]
Unfortunately a bug has been found. Run test `` with the following table settings:
```
SETTINGS min_bytes_for_wide_part = 1000000000, index_granularity = 8192, index_granularity_bytes = '10Mi', ttl_only_drop_parts = 1
```

It results in:
```
Received exception from server (version 24.12.1):
Code: 341. DB::Exception: Received from localhost:9000. DB::Exception: Exception happened during execution of mutation 'mutation_4.txt' with part 'all_1_2_1_3' reason: 'Code: 10. DB::Exception: Not found column id: in block event_time DateTime UInt32(size = 0). (NOT_FOUND_COLUMN_IN_BLOCK) (version 24.12.1.1)'. This error maybe retryable or not. In case of unretryable error, mutation can be killed with KILL MUTATION query 

2024.11.28 16:52:47.480306 [ 749691 ] {} <Error> MergeTreeBackgroundExecutor: Exception while executing background task {80f2698b-d321-4774-aa9b-4f0af748af14::all_1_2_1_4}: Code: 10. DB::Exception: Not found column id: in block event_ti
me DateTime UInt32(size = 0). (NOT_FOUND_COLUMN_IN_BLOCK), Stack trace (when copying this message, always include the lines below):

0. ./contrib/llvm-project/libcxx/include/exception:141: Poco::Exception::Exception(String const&, int) @ 0x0000000014fc5f32
1. ./build_clang/./src/Common/Exception.cpp:109: DB::Exception::Exception(DB::Exception::MessageMasked&&, int, bool) @ 0x000000000bc052b9
2. DB::Exception::Exception(PreformattedMessage&&, int) @ 0x0000000006ac4f0c
3. DB::Exception::Exception<String const&>(int, FormatStringHelperImpl<std::type_identity<String const&>::type>, String const&) @ 0x0000000006adc66b
4. ./build_clang/./src/Interpreters/ActionsDAG.cpp:899: DB::ActionsDAG::evaluatePartialResult(std::unordered_map<DB::ActionsDAG::Node const*, DB::ColumnWithTypeAndName, std::hash<DB::ActionsDAG::Node const*>, std::equal_to<DB::ActionsDA
G::Node const*>, std::allocator<std::pair<DB::ActionsDAG::Node const* const, DB::ColumnWithTypeAndName>>>&, std::vector<DB::ActionsDAG::Node const*, std::allocator<DB::ActionsDAG::Node const*>> const&, unsigned long, bool) @ 0x000000001
0566adf
5. ./build_clang/./src/Interpreters/ActionsDAG.cpp:814: DB::ActionsDAG::updateHeader(DB::Block const&) const @ 0x0000000010564f10
6. ./build_clang/./src/Processors/Transforms/ExpressionTransform.cpp:10: DB::ExpressionTransform::ExpressionTransform(DB::Block const&, std::shared_ptr<DB::ExpressionActions>) @ 0x0000000012b39630
7. ./contrib/llvm-project/libcxx/include/__memory/construct_at.h:35: DB::MutateAllPartColumnsTask::prepare() @ 0x00000000124ae143
8. ./build_clang/./src/Storages/MergeTree/MutateTask.cpp:1372: DB::MutateAllPartColumnsTask::executeStep() @ 0x00000000124ab85e
9. ./build_clang/./src/Storages/MergeTree/MutateTask.cpp:1982: DB::ExecutableTaskDropTTLExpiredPartsDecorator::executeStep() @ 0x00000000124b0276
10. ./build_clang/./src/Storages/MergeTree/MutateTask.cpp:2092: DB::MutateTask::execute() @ 0x0000000012496bc2
11. ./build_clang/./src/Storages/MergeTree/MutatePlainMergeTreeTask.cpp:99: DB::MutatePlainMergeTreeTask::executeStep() @ 0x00000000125f80ba
12. ./build_clang/./src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp:308: DB::MergeTreeBackgroundExecutor<DB::DynamicRuntimeQueue>::routine(std::shared_ptr<DB::TaskRuntimeData>) @ 0x000000001226ab2f
13. ./build_clang/./src/Storages/MergeTree/MergeTreeBackgroundExecutor.cpp:353: DB::MergeTreeBackgroundExecutor<DB::DynamicRuntimeQueue>::threadFunction() @ 0x000000001226b1e6
14. ./contrib/llvm-project/libcxx/include/__functional/function.h:848: ? @ 0x000000000bcd568e
15. ./contrib/llvm-project/libcxx/include/__functional/invoke.h:359: ? @ 0x000000000bcda002
16. ./contrib/llvm-project/libcxx/include/__functional/function.h:848: ? @ 0x000000000bcd3183
17. ./contrib/llvm-project/libcxx/include/__functional/invoke.h:359: ? @ 0x000000000bcd845b
18. ? @ 0x00007ba47fa94ac3
19. ? @ 0x00007ba47fb26850
 (version 24.12.1.1)
```
I'll revert for now.

**Comment by zvonand** [comment]
> Unfortunately a bug has been found. Run test `` with the following table settings:

Which test?

**Comment by zvonand** [comment]
@CurtizJ you did not specify the test

**Comment by CurtizJ** [comment]
Sorry, test `02521_lightweight_delete_and_ttl.sql`.

**Comment by zvonand** [comment]
Ok, I will take a look and reintroduce it

**Comment by zvonand** [comment]
@CurtizJ could you please share the full log of the failed test run? I do not see this exact fail in ci db

**Pull request merged**
